src/api/createDeviceAccreditedSystem/src/v1/steps.py

```diff
@@ -0,0 +1,178 @@
+from domain.repository.questionnaire_repository import (
+    QuestionnaireInstance,
+    QuestionnaireRepository,
+)
+from domain.request_models import CpmProductPathParams, CreateAsDeviceIncomingParams
+from domain.response.validation_errors import mark_validation_errors_as_inbound
+
+
+@mark_validation_errors_as_inbound
+def parse_as_device_payload(data, cache) -> CreateAsDeviceIncomingParams:
+    payload: dict = data[parse_event_body]
+    return CreateAsDeviceIncomingParams(**payload)
+
+
+def read_device_reference_data(data, cache) -> list[DeviceReferenceData]:
+    path_params: CpmProductPathParams = data[parse_path_params]
+    product_repo = DeviceReferenceDataRepository(
+        table_name=cache["DYNAMODB_TABLE"], dynamodb_client=cache["DYNAMODB_CLIENT"]
+    )
+    device_reference_data = product_repo.search(
+        product_id=path_params.product_id,
+        product_team_id=path_params.product_team_id,
+    )
+
+    # Only 1 AS DRD and only 1 MHS DRD
+
+    if not device_reference_data or len(device_reference_data) < 2:
+        raise ConfigurationError(
+            "You must configure the AS and MessageSet Device Reference Data before creating an AS Device"
+        )
+
+    if len(device_reference_data) > 2:
+        raise AccreditedSystemFatalError(
+            "More that 2 MessageSet Device Reference Data resources were found. This is not allowed"
+        )
+
+    exists = set()
+    if any(drd.name in exists or exists.add(drd.name) for drd in device_reference_data):
+        raise ConfigurationError(
+            "Only one AS and one MessageSet Device Reference Data is allowed before creating an AS Device"
+        )
+
+    return device_reference_data
+
+
+def read_spine_as_questionnaire(data, cache) -> Questionnaire:
+    return QuestionnaireRepository().read(QuestionnaireInstance.SPINE_AS)
+
+
+def validate_spine_as_questionnaire_response(data, cache) -> QuestionnaireResponse:
+    spine_as_questionnaire: Questionnaire = data[read_spine_as_questionnaire]
+    payload: CreateAsDeviceIncomingParams = data[parse_as_device_payload]
+    spine_as_questionnaire_response = payload.questionnaire_responses[
+        QuestionnaireInstance.SPINE_AS
+    ]
+    return spine_as_questionnaire.validate(
+        data=spine_as_questionnaire_response.__root__[0]
+    )
+
+
+def create_as_device(data, cache) -> Device:
+    product: CpmProduct = data[read_product]
+    payload: CreateAsDeviceIncomingParams = data[parse_as_device_payload]
+
+    # Create a new Device dictionary excluding 'questionnaire_responses'
+    device_payload = payload.dict(exclude={"questionnaire_responses"})
+    return product.create_device(**device_payload)
+
+
+def create_party_key_tag(data, cache) -> DeviceTagAddedEvent:
+    as_device: Device = data[create_as_device]
+    return as_device.add_tag(party_key=data[get_party_key])
+
+
+def create_device_keys(data, cache) -> Device:
+    # We will need to add some keys in the future, ASID?
```

Comment:
Did we resolve this/capture this in another ticket somewhere? I can't remember if we said ASID's were tags or not?

Reply:
We agreed that we would add it in another ticket. https://nhsd-jira.digital.nhs.uk/browse/PI-666